cvmfs/cache_ram.cc

```diff
@@ -141,59 +206,156 @@ void RamCacheManager::CtrlTxn(
   const int flags,
   void *txn)
 {
+  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
+  transaction->description = description;
+  transaction->object_type = type;
+  LogCvmfs(kLogCache, kLogDebug, "modified transaction %s",
+           transaction->id.ToString().c_str());
+  perf::Inc(counters_.n_ctrltxn);
 }
 
 
 int64_t RamCacheManager::Write(const void *buf, uint64_t size, void *txn) {
-  return -EIO;
+  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
+
+  if (transaction->pos + size > transaction->size) {
+    if (transaction->expected_size == kSizeUnknown) {
+      perf::Inc(counters_.n_realloc);
+      transaction->size = max(2*transaction->size, size + transaction->pos);
+      LogCvmfs(kLogCache, kLogDebug, "reallocate transaction for %s to %u B",
+               transaction->id.ToString().c_str(), transaction->size);
+      transaction->buffer = realloc(transaction->buffer, transaction->size);
+      if (!transaction->buffer) {
+        LogCvmfs(kLogCache, kLogDebug, "realloc failed: %s", strerror(errno));
+        return -errno;
+      }
+    } else {
+      LogCvmfs(kLogCache, kLogDebug,
+               "attempted to write more than requested (%u>%u)",
+               size, transaction->size);
+      return -ENOSPC;
+    }
+  }
+
+
+  uint64_t copy_size = min(size, transaction->size - transaction->pos);
```

Comment:
At this point, we already know that the remaining buffer is large enough to host `size` bytes, no?